### PR TITLE
Contributor documentation - updating components

### PR DESCRIPTION
**Pull request by pouretrebelle**
Part of https://github.com/github/primer/issues/534

https://github.com/primer/view_components/blob/pouretrebelle/docs-updating-components/contributor-docs/updating-components.md

## Contents

- [x] Deprecation and removal process
- [x] Workflow with Primer CSS changes
  Dependent on https://github.com/primer/view_components/pull/970

### CODE REVIEW

File: contributor-docs/updating-components.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+   For significant changes consider the [component lifecycle](https://primer.style/contribute/component-lifecycle) criteria; if the changes qualify the component for the next maturity level it’s much easier to introduce, for example, a new Beta component and deprecate the Alpha component than maintaining both APIs in a single component. This also makes it easier to identify which cases have been migrated.
+
+2. Mark old functionality as deprecated
+
+   For arguments add `DEPRECATED` before the YARD argument description, for components update the `status` value to `:deprecated`.
+
+3. Write an ERB linter to detect and fix uses of old functionality
+
+   This will make it easier to find uses of the old functionality in github.com. Our ERB linters also have autocorrection logic so we can automate migrations where possible. [See our linting guide for details on how to write linters](https://primer.style/view-components/linting).
```
Comment (pouretrebelle) [inline contributor-docs/updating-components.md:19]: Link to be updated to internal docs when written

File: contributor-docs/updating-components.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+4. Make a release and upgrade the Primer ViewComponents library in github.com
+5. Migrate uses in github.com to new functionality
+
+   [How do we run the migrations / get data from the linter?]
```
Comment (pouretrebelle) [inline contributor-docs/updating-components.md:24]: @manuelpuyol can you write a bit up on this? I remember going through it in a call before xmas but can't find my notes :(

File: contributor-docs/updating-components.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+   If the migrations result in a lot of changes it’s best to split them up into multiple pull requests. [Is there a strategy for this to ping as few teams as possible?]
```
Comment (pouretrebelle) [inline contributor-docs/updating-components.md:26]: @jonrohan I remember you doing this for some css vars migration. Did you use some sort of script for it? 🙏

File: contributor-docs/updating-components.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+6. Remove old functionality
+
+   Deprecated arguments can be removed as soon as they’re no longer used in github.com. If the status has changed the old component can be removed after one month [TBC].
```
Comment (pouretrebelle) [inline contributor-docs/updating-components.md:30]: [Slack thread about this](https://github.slack.com/archives/CSGAVNZ19/p1642085899035000)

File: contributor-docs/updating-components.md (hunk 3 above)
Comment (jonrohan) [reply contributor-docs/updating-components.md:26]: Yes, I did with this script https://thehub.github.com/engineering/products-and-services/dotcom/serviceowners/serviceowners-usage/#bincodeowners-stash-changes

Then I scripted that script with this to automatically publish draft prs for me:

```rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Get all currently modified files.
modified_files = `git status --short`.lines
                  .select { |line| line.start_with?(/^\s?M\s+/) }
                  .map { |line| line.strip.sub(/^M\s+/, "") }

# Get list of teams that own the modified files.
teams = `bin/codeowners #{modified_files.join(" ")}`.gsub("@github/","").split("\n")

teams.each do |team|
  full_team_name = "@github/#{team}"
  branch_name = "migrating_deprecated_system_args/#{team}"

  # checkout master
  `git checkout master`
  # stash pop
  `git stash pop`
  # create team branch
  `git checkout -b #{branch_name}`
  # run script
  `bin/codeowners-stash-changes #{full_team_name}`
  # commit changes
  `git commit -am "Migrating deprecated primer system arguments"`
  `git push -u origin #{branch_name}`
  # push draft pr
  `gh pr create -d -b "" -t "[Color modes: v2 migration] Migrating deprecated Primer View Component system arguments in github/#{team} owned files"`
end
```

Comment (pouretrebelle) [reply contributor-docs/updating-components.md:26]: 😍  a thing of beauty

File: contributor-docs/updating-components.md (hunk 2 above)
Comment (manuelpuyol) [reply contributor-docs/updating-components.md:24]: You can either enable your linter globally adding it to the [`.erb-lint.yml`](https://github.com/github/github/blob/master/.erb-lint.yml) file and running `bin/bundle exec erblint -a app/views app/components app/packages`

or you can run only your linter with `bin/bundle exec erblint --enable-linters your_linter_name -a app/views app/components app/packages`

File: contributor-docs/updating-components.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+4. Make a release and upgrade the Primer ViewComponents library in github.com
+5. Migrate uses in github.com to new functionality
+
+   To use your linter run:
+
+   ```bash
+   bin/bundle exec erblint --enable-linters LinterName -a app/views app/components app/packages
+   ```
+
+   You can also enable your linter globally by adding it to `[.erb-lint.yml](https://github.com/github/github/blob/master/.erb-lint.yml)`.
```
Comment (pouretrebelle) [inline contributor-docs/updating-components.md:30]: I'm not sure of the cases when we'd want to add it as a global linter and when not. I presume autocorrection ones can't be global bc how would it commit the changes ?

**Review by jonrohan — APPROVED** [review]
Nice work 👍🏻

**Pull request merged**